Porting HDF5 recorder environment along with Pinocchio Solver

zephirefaith opened the pull request:
## Motivation and Context

Solves Issue #60 

Ports over environments used for collecting and replaying demonstrations on Stretch. Also ports in Pinocchio IK solver which fulfills immediate need for manipulation motions. Since SLAP use-case is pretty similar to SANER's I am assuming this is the best way to attack the problem.

## Changelog

<!--- Itemized list of changes -->

## How Has This Been Tested

<!--- Please describe here how your modifications have been tested. -->

## Types of changes

<!-- What types of changes does your code introduce? -->
<!-- Put an `x` in all the boxes that apply: -->
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

cpaxton commented on projects/collect_h5.py:
```diff
@@ -0,0 +1,72 @@
+    and time.
+    """
+
+    def __init__(self, task_name, dir_path):
```
type hints

cpaxton replied:
might be good to document what these 2 are as well

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+    DT = 1e-1
+    DAMP = 1e-12
+
+    def __init__(self, urdf_path, ee_link_name, controlled_joints):
```
type hints/descriptions

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+]
+
+
+class PinocchioIKSolver:
```
1 sentece class description

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+        """returns dof for the manipulation chain"""
+        return len(self.controlled_joints)
+
+    def _qmap_control2model(self, q_input):
```
hints

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+
+        return q_out
+
+    def _qmap_model2control(self, q_input):
```
hints

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+        pinocchio.forwardKinematics(self.model, self.data, q_model)
+        pinocchio.updateFramePlacement(self.model, self.data, self.ee_frame_idx)
+        pos = self.data.oMf[self.ee_frame_idx].translation
+        quat = R.from_matrix(self.data.oMf[self.ee_frame_idx].rotation).as_quat()
```
should return type be (pos, quat) like this, or are we using sophus?
cc @exhaustin 

for now best to be consistent with pybullet i guess

zephirefaith replied:
What is "consistent with pybullet"?

exhaustin replied:
[pos, quat] is the pybullet standard: https://usermanual.wiki/Document/pybullet20quickstart20guide.479068914/html#pf17

I think we should unify pose representations across the entire repo but I don't have a very good candidate right now. Sophus is nice but it needs pybind to install which is slightly annoying. Maybe we can wait for the Theseus Lie library to release. Anyhow I think it's okay to just use pos, quat for now.

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+        i = 0
+        q = self.q_neutral.copy()
+        desired_ee_pose = pinocchio.SE3(R.from_quat(quat).as_matrix(), pos)
+        while True:
```
i don't like "while True", should probably be a max numebr of iter?

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+            if np.linalg.norm(err) < self.EPS:
+                success = True
+                break
+            if i >= max_iterations:
```
oh i see, its here

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+
+
+class CEM:
+    def __init__(self, max_iterations, num_samples, num_top, tol):
```
hints + descriptions

cpaxton commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,246 @@
+        return cost_arr[i_best], aux_outputs[i_best]
+
+
+if __name__ == "__main__":
```
could you make this into a short test??

cpaxton replied:
like in test/.../motions

cpaxton commented on src/home_robot/home_robot/agent/motion/stretch.py:
```diff
@@ -179,14 +180,28 @@ def _create_ik_solvers(self):
+                self.manip_mode_urdf_path,
+                self.ee_link_name,
+                self.manip_mode_controlled_joints,
+            )
```
else - ik_type not recongized runtime error

cpaxton commented on src/home_robot/home_robot/agent/motion/stretch.py:
```diff
@@ -179,14 +180,28 @@ def _create_ik_solvers(self):
+        urdf_path=None,
+        visualize=False,
+        root=".",
+        ik_type: str = "pybullet",
```
hints please

cpaxton commented on src/home_robot/home_robot/agent/motion/stretch.py:
```diff
@@ -435,14 +451,19 @@ def get_link_pose(self, link_name, q=None):
         return pose
 
-    def update_head(self, qi, look_at):
+    def update_head(self, qi, look_at) -> np.ndarray:
```
qi: np.ndarray

cpaxton commented on src/home_robot/home_robot/utils/data_tools/writer.py:
```diff
@@ -91,7 +91,7 @@ def fix_data(self, data):
         return new_data
 
-    def flatten_dict(self, key, data):
+    def flatten_dict(self, key, data) -> Dict:
```
hard to do types here right? maybe we dont need hints. but key = str

zephirefaith replied:
Yeah the values range from str, Dict, Lists of ndarrays. I will add key types only since that is consistent.

cpaxton reviewed (commented):
a few minor documentation fixes + maybe we should copy your testing logic into the tests folder?

cpaxton commented on tests/home_robot/motion/test_pinocchio_ik.py:
```diff
@@ -0,0 +1,56 @@
+    PinocchioIKSolver)
+
+if __name__ == "__main__":
+    pos_desired = np.array([-0.10281811, -0.7189281, 0.71703106])
```
we should make this an actual test, like the other IK test -

cpaxton reviewed (commented):
just one last thing, just check the inputs + outputs (vs forward kinematics) for your test to make sure its really working

zephirefaith commented:
@exhaustin I see this branch failing the tests on CI. I looked at the error message and it is due to the inability to find `rospkg`. Where can I add that requirement so we don't see this superfluous failure?

cpaxton reviewed (approved):
lgtm once tests are passing

exhaustin reviewed (changes requested):
Sorry for the late review. 

As the original author of the Pinocchio IK solver and CEM implementation, I'm happy to help make changes in those places, feel free to delegate stuff to me.

zephirefaith commented on src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:
```diff
@@ -0,0 +1,221 @@
+        num_top = 10  # TODO: what is this?
+        pos_error_tol = 0.005
+        ori_error_tol = 0.2
+        pos_desired, quat_desired = pose_query
+        ik_solver = self
+        pos_wt = 1.0
+        rot_wt = 0.0
+
+        opt = CEM(
+            max_iterations=max_iterations,
+            num_samples=num_samples,
+            num_top=num_top,
+            tol=pos_error_tol,
+        )
+
+        def solve_ik(dr):
+            pos = pos_desired
+            quat = (R.from_rotvec(dr) * R.from_quat(quat_desired)).as_quat()
+
+            q, _ = ik_solver.compute_ik(pos, quat)
+            pos_out, rot_out = ik_solver.compute_fk(q)
+
+            cost_pos = np.linalg.norm(pos - pos_out)
+            cost_rot = (
+                1 - (rot_out * quat_desired).sum() ** 2
+            )  # TODO: just minimize dr?
+
+            cost = pos_wt * cost_pos + rot_wt * cost_rot
+
+            return cost, q
+
+        cost_opt, q_result = opt.optimize(
+            solve_ik, x0=np.zeros(3), sigma0=np.array([0, 0, ori_error_tol / 2])
+        )
+        pos_out, quat_out = ik_solver.compute_fk(q_result)
+        print(
+            f"After ik optimization, cost: {cost_opt}, result: {pos_out, quat_out} vs desired: {pose_query}"
+        )
+        # pos_out, quat_out = self.fk(q_result)
+        return q_result
```
Could you take this task? Along with making this solution more generic?

exhaustin replied:
I just made the changes and also updated the tests, let me know if they look good!

zephirefaith commented on tests/home_robot/motion/test_pinocchio_ik.py:
```diff
@@ -0,0 +1,76 @@
+    assert (
+        best_cost <= POS_ERROR_TOL
+        or np.all(opt_sigma) <= POS_ERROR_TOL
+        or last_iter >= CEM_MAX_ITERATIONS
```
My reasoning wasn't to test if it ended early but that only asserting for error to be < tolerances provided is too tight a test. Thus to check full functionality I added checking for "best effort" optimization by checking `last_iter >= max_iter`

exhaustin replied:
Sorry I don't follow. Are you're saying that you know that the problem you're giving to the optimizer is impossible, so you expect the optimizer to run until max iters as a best effort?

zephirefaith commented:
@exhaustin I am assuming the indicator_block in the tests is for visualization and debugging purposes? That's my assumption on the decision to replicate it for both `pin_robot` and `pb_robot` for checking their FK+IK consistency against each other.

The basic logic is to take your `test_ik()` from `test_ik.py` but run those poses against both `pin_robot` and `pb_robot`. Then compute error/distance b/w the resultant `pos, quat` from both solvers to check for consistency between the solvers.

exhaustin commented:
> @exhaustin I am assuming the indicator_block in the tests is for visualization and debugging purposes? That's my assumption on the decision to replicate it for both `pin_robot` and `pb_robot` for checking their FK+IK consistency against each other.
> 
> The basic logic is to take your `test_ik()` from `test_ik.py` but run those poses against both `pin_robot` and `pb_robot`. Then compute error/distance b/w the resultant `pos, quat` from both solvers to check for consistency between the solvers.
> 
> Also, if you could take up the task to factor out current optimization-based solver to `UnderactuatedIKSolver` and make the current implementation more generic then that would be super-helpful! Thanks.

- Yes the block is for visualization, a nice addition from Chris.
- I just factored out the IK optimizer, let me know if it looks good.
- Now that the tests have more shared components and are more complicated, it'd be nice to use some [pytest.fixtures](https://docs.pytest.org/en/6.2.x/fixture.html). I can do it later if you don't have time, lmk.

zephirefaith commented:
@exhaustin Changes look good to me. I like factoring out the optimizer as its own solver-agnostic class. Thanks!

I fixed the conflict and moved out `pinocchio_ik_solver.py` to the new motion folder to respect the new organization. Once the test passes let's merge this?

exhaustin reviewed (approved):
Approving now.
@zephirefaith feel free to leav comments on my changes, or just merge if my changes look good.

zephirefaith commented:
Thanks for helping make the tests nice and modular @exhaustin 

Changes look good and the tests pass which is in itself a good marker!
